File: ch3/barrier/golang/barrier2.go
package main

import (
	"fmt"
	"sync"
	"context"
	"golang.org/x/sync/semaphore"
	// "time"
)

type Counter struct {
	val int
	m sync.Mutex
}

func main() {
	numThreads := 10      // Number of threads to create
	count := Counter{val: 0}
	ctx := context.TODO() // Context req for golang's semaphores

	turnstile1 := semaphore.NewWeighted(int64(1))
	turnstile2 := semaphore.NewWeighted(int64(1))
	// Turnstile2 : bomb needs to be detonated
	turnstile1.Acquire(ctx, 1)

	done := make(chan int, numThreads)
	for i := 0; i < numThreads; i++ {
		go func() {

			//--------------------------------------------------------
			fmt.Println("rendezvous1")

			// First barrier
			count.m.Lock()
			count.val += 1
			if count.val == numThreads {
				turnstile2.Acquire(ctx,1)  // diffuse bomb2
				turnstile1.Release(1)      // activate bomb1
			}
			count.m.Unlock()
			turnstile1.Acquire(ctx, 1)
			turnstile1.Release(1)

			//--------------------------------------------------------
			fmt.Println("critical point1")

			// Second barrier
			count.m.Lock()
			count.val -= 1
			if count.val == 0 {
				turnstile1.Acquire(ctx,1)  // diffuse bomb1
				turnstile2.Release(1)      // activate bomb2
			}
			count.m.Unlock()
			turnstile2.Acquire(ctx, 1)
			turnstile2.Release(1)

			//--------------------------------------------------------
			fmt.Println("rendezvous2")

			// First barrier
			count.m.Lock()
			count.val += 1
			if count.val == numThreads {
				turnstile2.Acquire(ctx,1)  // diffuse bomb2
				turnstile1.Release(1)      // activate bomb1
			}
			count.m.Unlock()
			turnstile1.Acquire(ctx, 1)
			turnstile1.Release(1)

			//--------------------------------------------------------
			fmt.Println("critical point2")

			// Second barrier
			count.m.Lock()
			count.val -= 1
			if count.val == 0 {
				turnstile1.Acquire(ctx,1)  // diffuse bomb1
				turnstile2.Release(1)      // activate bomb2
			}
			count.m.Unlock()
			turnstile2.Acquire(ctx, 1)
			turnstile2.Release(1)

			done <- 1
		}()
	}

	for i := 0; i < numThreads; i++ {
		<-done
	}
}

File: ch3/barrier/java/Barrier2.java
import java.util.concurrent.Semaphore;

public class Barrier {

    class MyThread extends Thread {
        Semaphore barrier;
        Semaphore mutex;
        int[] count;
        int n;
        MyThread(Semaphore b, Semaphore m, int[] count, int numOfThreads) {
            barrier = b;
            mutex = m;
            this.count = count;
            n = numOfThreads;
        }
        public void run() {
            System.out.println("Statement #1 for thread " + this.getName());
            try {
                mutex.acquire();
            }
            catch (Exception e) {
                System.out.println("mutex.acquire exception " + e + " in thread " + this.getName());
            }
            count[0] = count[0] + 1;
            if (count[0] == n) {
                barrier.release();
            }
            mutex.release();
            System.out.println("mutex released by thread " + this.getName());
            try {
                barrier.acquire();
            }
            catch (Exception e) {
                System.out.println("barrier.acquire exception " + e + " in thread " + this.getName());
            }
            barrier.release();
            System.out.println("Statement #2 for thread " + this.getName());
        }
    }

    public static void main(String[] args) {
        Barrier s = new Barrier();
        int numOfThreads = 7;
        int[] count = {0};
        Semaphore allArrived = new Semaphore(0);
        Semaphore mutex = new Semaphore(1);
        MyThread[] threads = new MyThread[numOfThreads];
        for (int i = 0; i < numOfThreads; i++) {
            MyThread a = s.new MyThread(allArrived, mutex, count, numOfThreads);
            threads[i] = a;
            a.start();
        }
        for (int i = 0; i < numOfThreads; i++) {
            try {
                threads[i].join();
            }
            catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

}
